### PR TITLE
[overlay] Port Foundation String extensions to StringProtocol

### PR DESCRIPTION
**Pull request by moiseev**

### CODE REVIEW

**Comment by moiseev** [comment]
@swift-ci Please test

**Comment by moiseev** [comment]
@swift-ci Please smoke benchmark

**Comment by moiseev** [comment]
@swift-ci Please Test Source Compatibility

**Comment by parkera** [comment]
I'm going to need some time to review this. Please hold off on merging.

**Comment by moiseev** [comment]
@parkera Sure thing. It is mostly copy-paste of 'living' methods from `extension String` to `extension StringProtocol where Index == String.Index` and unavailable ones from `extension String` to `extension StringProtocol` (and removing the bodies, since they are unavailable anyway).

**Comment by moiseev** [comment]
Oh, one more thing: the initializers are kept in `extension String`.

**Comment by moiseev** [comment]
Compatibility breakage looks unrelated.

**Comment by moiseev** [comment]
@swift-ci Please Test Source Compatibility

File: stdlib/public/SDK/Foundation/NSStringAPI.swift [hunk 1]
```diff
@@ -230,849 +170,1202 @@ extension String {
...
-  public init(
-    contentsOf url: URL
-  ) throws {
-    let ns = try NSString(contentsOf: url, usedEncoding: nil)
-    self = ns as String
+  // @property NSStringEncoding fastestEncoding;
+
+  /// The fastest encoding to which the string can be converted without loss
+  /// of information.
+  public var fastestEncoding: String.Encoding {
```
Comment (phausler) [inline stdlib/public/SDK/Foundation/NSStringAPI.swift]: we should be able to calculate this without bridging

File: stdlib/public/SDK/Foundation/NSStringAPI.swift [hunk 2]
```diff
@@ -230,849 +170,1202 @@ extension String {
...
+    options opts: NSLinguisticTagger.Options = [],
+    orthography: NSOrthography? = nil,
+    invoking body:
+      (String, Range<Index>, Range<Index>, inout Bool) -> Void
+  ) {
+    _ns.enumerateLinguisticTags(
+      in: _toNSRange(range),
+      scheme: tagScheme,
+      options: opts,
+      orthogr
```
Comment (phausler) [inline stdlib/public/SDK/Foundation/NSStringAPI.swift]: it is a bit of a shame that enumerateSubstrings enumerates Strings and not Substrings...

Comment (phausler) [inline stdlib/public/SDK/Foundation/NSStringAPI.swift]: is this really the case? the performance behavior is different iirc and does the same UTF8 to backing storage happen too?

File: stdlib/public/SDK/Foundation/NSStringAPI.swift [hunk 3]
```diff
@@ -1602,140 +1728,17 @@ extension String {
...
-    return r
-  }
 }
 
 // Pre-Swift-3 method names
-extension String {
+extension StringProtocol {
 
   @available(*, unavailable, renamed: "localizedName(of:)")
   public static func localizedNameOfStringEncoding(
```
Comment (phausler) [inline stdlib/public/SDK/Foundation/NSStringAPI.swift]: why are we moving them to StringProtocol for static functions?

**Review by phausler — COMMENTED** [review]
I still need to go over this a bit more but as a cursory review I have a couple of questions:

1) In a lot of places we take another String as a parameter, can we perhaps make that be generic and take StringProtocol? A number of the methods would be much more useful if that was the case

2) Can we adopt a consistent indentation and line breaking to the rest of the Foundation overlay since this is effectively a re-write?

3) I think we can do a lot better for bridging here for the _ns accessor. In almost every case we know the NSString will never be stored past the invocation of the method so we could use noCopy variants to get a nice perf-boost.

4) In the cases we take Range, should we not also consider widening the net and take RangeExpression as a generic argument?

**Comment by moiseev** [comment]
@swift-ci Please test

**Comment by moiseev** [comment]
@swift-ci Please Test Source Compatibility

**Comment by moiseev** [comment]
@swift-ci Please test

**Comment by moiseev** [comment]
@swift-ci Please smoke benchmark

**Comment by moiseev** [comment]
@swift-ci Please Test Source Compatibility

**Comment by moiseev** [comment]
@swift-ci Please Test Source Compatibility

**Comment by moiseev** [comment]
@swift-ci Please smoke benchmark

**Comment by moiseev** [comment]
Compatibility test suite failure is a UPASS.

**Comment by moiseev** [comment]
@swift-ci Please Test Source Compatibility

**Comment by phausler** [comment]
@swift-ci please smoke benchmark

**Comment by phausler** [comment]
I think we are good to go here. I will revise my performance changes to reflect the protocol migration changes

**Comment by phausler** [comment]
Good news is that the changes I have in mind to boost the performance will counteract any losses we have here

**Comment by moiseev** [comment]
Just wanted to ask you about it. Thanks! I'm merging, then.

**Pull request merged**